packages/openactive-integration-tests/test-data-generator/README.md

```diff
@@ -0,0 +1,78 @@
+# test-data-generator
+
+Command line tool to generate a distribution definition of test data based on the configured features.
+
+Test data is outputted to `./test-data/test-data.json` by default.
+
+## Usage
+
+```bash
+npm run test-data-generator # all active features specified in the current config
+
+npm run test-data-generator authentication # subset of active features within the authentication category
+
+npm run test-data-generator common-error-conditions # only enough data for the common-error-conditions feature
+
+npm run test-data-generator --output ./test-data.json # output to ./test-data.json
+```
+
+## Test data file format
+
+The `item` within the file format is identical to the Controlled Opportunity Creation endpoint defined within the [OpenActive Test Interface](https://openactive.io/test-interface/). This allows an implementation to use the same logic to generate data from the test data file and from the test interface.
+
+```json
+{
+  "@context": [
+    "https://openactive.io/",
+    "https://openactive.io/test-interface"
+  ],
+  "@type": "ItemList",
+  "numberOfItems": 123,
+  "itemListElement": [
+    {
+      "@type": "ListItem",
+      "item": {
+        "@type": "ScheduledSession",
+        "superEvent": {
+          "@type": "SessionSeries",
+          "organizer": {
+            "@type": "Organization",
+            "@id": "https://id.booking-system.example.com/organizer/3"
+          }
+        },
+        "test:testOpportunityCriteria": "https://openactive.io/test-interface#TestOpportunityBookable",
+        "test:startDateMin": "2020-12-20",
+        "test:startDateMax": "2020-12-25"
```

Comment:
Just on this part of the model, I wonder whether we should have the requirements within a single type - given this is how they're likely processed, and also given that the requirements include a mix of opportunityRequirements and orderRequirements, so perhaps don't make sense on the Event itself, so e.g.

```suggestion
        "test:testDataRequirements": {
          "@type": "test:TestDataRequirements",
          "test:startDateMin": "2020-12-20",
          "test:startDateMax": "2020-12-25"
        }
```

Alternatively perhaps we create a separation between opportunity and offer requirements...?

```suggestion
        "test:testOpportunityDataRequirements": {
          "@type": "test:OpportunityTestDataRequirements",
          "test:startDateMin": "2020-12-20",
          "test:startDateMax": "2020-12-25"
        }
        "test:testOfferDataRequirements": {
          "@type": "test:OfferTestDataRequirements",
          "test:availableChannelIncludes": [
            "availableChannelIncludes"
          ]
        }
```

Reply:
@nickevansuk I agree with both suggestions.

Also, I wonder if we should structure the fields themselves? There's turning out to be a lot of different fields. It would be easier to implement (for external Booking Systems) if we abstracted the commonalities.

I've given a rough idea of what this might look like here at the top of TestDataRequirements.d.ts (and you can also see how many different fields there are so far. I've extended your suffix scheme for organising them)

If the fields are unstructured, a BS's logic might look like (for a controlled "create opportunity" request):

```js
const prepayment = generateChoice(PREPAYMENT_OPTIONS, {
  allowlist: testDataRequirements.prepaymentAllowlist,
  blocklist: null,
  allowNull: testDataRequirements.prepaymentAllowNull,
});
const eventStatus = generateChoice(EVENT_STATUS_OPTIONS, {
  allowlist: null,
  blocklist: testDataRequirements.eventStatusBlocklist,
  allowNull: null
});
```

EW. And what if we eventually add a `prepaymentBlocklist` field - they'd have to add it (or more likely - not realise and start generating incorrect data).

If the fields are structured, the same BS's logic might look like:

```js
const prepayment = generateField(testDataRequirements.prepayment);
const eventStatus = generateField(testDataRequirements.eventStatus);
```

The field could look like:

```js
prepayment: {
  type: 'Allowlist',
  mustBeOneOf: ['https://openactive.io/Unavailable'],
  valueType: 'oa:RequiredStatusType',
}
```

Therefore, the BS's `generateField(..)` function could generate everything it needs to from that definition, and would allow for adding new criteria or updating existing criteria without breaking existing BSs (unless we add a new type of requirement)

Reply:
Love it!